Improve Story block media loading

Tug opened the pull request:
Addresses part of https://github.com/Automattic/jetpack/issues/16591:
- Show a progress indicator when uploading media from the block editor, prevent story playback while its uploading
- Improve loading of media files
- Better fit media on the player, use heuristics to crop some part of the image to make it fit a user screen (for instance allow 10% cropping max, after which margins can be displayed)
- Refine UX
- Scale down story embeds (make it 50% smaller?)
 
With this change we stop rendering all the slides dynamically from the server. Instead we only render the first one and serialize the media into the markup so the client can retrieve it and load slides dynamically. Slides are rendered as the story is played, with a preload of the next slide.

It also applies the latest changes in design regarding stories:
- Smaller embeds
- Shadow and animation on hover
- Play and replay button removed, replaced with transparent overlay
- Displaying an expand icon on the top right

This doesn't update buttons to use `@wordpress/element` nor fixes accessibility issues, those are tackled in this PR https://github.com/Automattic/jetpack/pull/16852

Here is how it looks like:

Frontend:
<img width="639" alt="Screenshot 2020-08-10 at 15 27 37" src="https://user-images.githubusercontent.com/230230/89787621-13729680-db1e-11ea-9e4e-fe33052c3310.png">

Block Editor:
<img width="591" alt="Screenshot 2020-08-10 at 15 27 08" src="https://user-images.githubusercontent.com/230230/89787652-1ff6ef00-db1e-11ea-96e9-c61567e192f0.png">

#### Changes proposed in this Pull Request:
In this PR we lazy load media while stories are being played. This way, we improve the initial page load speed.
We also define a list of alternative images for each slide using srcset and force the browser to load the right image for the current viewport.

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
Load a page with a story in it. Notice in the network tab only the fist slide is loaded.
Disable JS and reload the page, notice the story looks almost identical as the version with JS

#### Proposed changelog entry for your changes:
-  Improve loading of media files for the Story block

Tug commented on extensions/blocks/story/story.php:
```diff
@@ -84,17 +163,97 @@ class="wp-story-video intrinsic-ignore wp-video-%3$s"
+ */
+function render_pagination_bullet( $slide_index ) {
+	return sprintf(
+		'<button class="wp-story-pagination-bullet" aria-label="Go to slide %d">
```
I didn't know, do you know more about this?
In general, I try to stay away from anchor tags when there is no `href` attribute to point to, which is the case here

jeherve replied:
This may tell you a bit more:
- p5TWut-gm-p2
- https://github.com/Automattic/wp-calypso/pull/31175 (that's back when Jetpack blocks were still developed in the Calypso monorepo, but the code still uses `a` tags to this day:
https://github.com/Automattic/jetpack/blob/3b5cf194ac4e43c57f07784d6f0171802314d3d9/extensions/blocks/slideshow/slideshow.js#L169

As you can see, the no href issue was discussed in that PR as well.

Tug replied:
Interesting, will convert every button, just to be on the safe side

Tug commented on extensions/blocks/story/player/components/calypso-spinner.js:
```diff
@@ -0,0 +1,14 @@
+ */
+import { createElement } from '@wordpress/element';
+
+const CalypsoSpinner = () => (
```
We explored this with Megs but she preferred we use the calypso spinner which looks much better for our case here.

Tug commented on extensions/blocks/story/story.php:
```diff
@@ -84,17 +163,97 @@ class="wp-story-video intrinsic-ignore wp-video-%3$s"
-function render_slide( $media, $index ) {
+function render_slide( $media, $index = 0 ) {
+	$media_template = '';
+	switch ( ! empty( $media['type'] ) && $media['type'] ) {
```
Thanks, indeed I updated this while you were reviewing 👍

Tug commented on extensions/blocks/story/story.php:
```diff
@@ -27,6 +31,58 @@ function register_block() {
 }
 add_action( 'init', __NAMESPACE__ . '\register_block' );
 
+/**
```
Good point. I think photon could simplify some of the logic here. I'm still not fully convinced this block should depend on jetpack features as it could actually be moved to gutenberg core or somewhere else. For now I think trying to simply reuse the already available versions of an image makes sense.

Tug commented:
Thanks for the review @jeherve I'm going to address your comments.

> But if I then add an additional image to that block, I lose the indicators:

Oh right I can reproduce when using drag n drop. Fixing this 👍 

> Things don't quite work on the frontend though:

Not sure how you ended up with this, it seems like nor the CSS nor the JS is loaded, do you have any error in the console?

jeherve commented:
> do you have any error in the console?

No error in the console, no. Not quite sure what happened. this may have been a fluke of my testing site, I'll try again in my next review.

Tug commented on extensions/blocks/story/story.php:
```diff
@@ -84,17 +163,97 @@ class="wp-story-video intrinsic-ignore wp-video-%3$s"
+function render_top_right_icon( $settings ) {
+	$show_slide_count = isset( $settings['showSlideCount'] ) ? $settings['showSlideCount'] : false;
+	$slide_count      = isset( $settings['slides'] ) ? count( $settings['slides'] ) : 0;
+	if ( $show_slide_count ) {
```
Actually I think it makes sense here, `$show_slide_count` is just a boolean that will decide if we render one type of icon or the other. `$slide_count` will be 0 if `$settings['slides']` is an empty array or if it's not set.

Tug commented:
@jeherve Thanks, I addressed most of your concerns. I only replaced the button elements on the server side for now. https://github.com/Automattic/jetpack/pull/16852 is supposed to replace the custom buttons with `@wordpress/components` ones and it needs updating so I'll investigate those as part of that PR.

Other than that, I'm not sure what your error is, are we talking about the site frontend, or the calypso reader (the latter is still in progress [here](https://github.com/Automattic/wp-calypso/pull/44985))? Could you double check that beta blocks enabled and that `story/view.css` and `story/view.js` load on the page? What browser are you using?

jeherve reviewed (approved):
This works well now, and I can't reproduce the issues I was facing earlier. 🚢
